client/allochealth/tracker.go

```diff
@@ -277,17 +277,37 @@ func (t *Tracker) watchTaskEvents() {
 		// Detect if the alloc is unhealthy or if all tasks have started yet
 		latestStartTime := time.Time{}
 		for taskName, state := range alloc.TaskStates {
-			// If the task is a poststop task we do not want to evaluate it
-			// since it will remain pending until the main task has finished
-			// or exited.
-			if t.lifecycleTasks[taskName] == structs.TaskLifecycleHookPoststop {
-				continue
-			}
+			if t.lifecycleTasks != nil && t.lifecycleTasks[taskName] != nil {
+				// If the task is a poststop task we do not want to evaluate it
+				// since it will remain pending until the main task has finished or
+				// exited.
+				if t.lifecycleTasks[taskName].Hook == structs.TaskLifecycleHookPoststop {
+					continue
+				}
 
-			// One of the tasks has failed so we can exit watching
-			if state.Failed || (!state.FinishedAt.IsZero() && t.lifecycleTasks[taskName] != structs.TaskLifecycleHookPrestart) {
-				t.setTaskHealth(false, true)
-				return
+				// If this is a poststart task which has already succeeded we
+				// want to check for two possible success conditions before
+				// attempting to evaluate it.
+				if t.lifecycleTasks[taskName].Hook == structs.TaskLifecycleHookPoststart && state.Successful() {
+
+					// If the task was successful and it's runtime is at least
+					// t.minHealthyTime, skip evaluation.
+					if state.FinishedAt.Sub(state.StartedAt) >= t.minHealthyTime {
+						continue
+					}
```

Comment:
This tracker has an existing special case for _pre_ start tasks, and I wonder if we should be using that here for poststart as well.

All tasks with lifecycle hooks without the `sidecar = true` field are supposed to be tolerated already as "short-lived ephemeral tasks" without any additional configuration:
* A sidecar prestart task needs to start up and keep running. If it exits for any reason, that's a failure.
* A non-sidecar prestart task needs to finish. It's only a failure if it exits with a non-0 exit code.

So rather than introducing a new `ignore_min_healthy_time` field, couldn't we use the same conditional here for poststart that we're using for prestart tasks? (I admit I could be missing something here for sure.)

Reply:
Ah, I think you're absolutely correct here. I've made the appropriate changes inside of the lifecycleTask loop and reverted my addition of `ignore_min_healthy_time`.

I did a little bit of testing and this has corrected both the 'not running for min_healthy_time' bug and the erroneous marking of previous allocations as unhealthy.

Reply:
> A non-sidecar prestart task needs to finish. It's only a failure if it exits with a non-0 exit code.

I updated the docstring for the `Successful` method of `structs.TaskState` to reflect this.